kernel/src/table_changes/resolve_dvs.rs

```diff
@@ -0,0 +1,420 @@
+            // if no deletion vector is given.
+            //  rm_dv  = Treemap_d(0, 1)
+            //  add_dv = Treemap_d(1, 2)
+            //
+            // The result of this commit is:
+            // - row 0 is restored
+            // - row 1 is unchanged
+            // - row 2 is deleted
+            // Thus for this commit we must generate `Treemap_s(0)` for the added rows, and
+            // `Treemap_s(2)` for deleted rows.
+            //
+            //  # Insertion Selection Treemap
+            //  The selection vector of added rows is calculated using set subtraction over deletion
+            //  treemaps `rm_dv - add_dv`. These rows went from set (deleted) in `rm_dv` to unset
+            //  (restored) in the `add_dv`. All other rows are either cancelled in set subtraction,
+            //  or were not selected in either treemap. Hence, they are not selected to
+            //  be in the set of changed rows. Applying this to our deletion treemaps:
+            //  rm_dv - add_dv =
+            //      Treemap_d(0, 1)
+            //    - Treemap_d(1, 2)
+            //    = Treemap_s(0)
+            //  The selection treemap shows that row 0 was inserted
+            //
+            //  # Deletion Selection Treemap
+            //  The selection vector of deleted rows is calculated using `add_dv - rm_dv`. These rows went
+            //  from unset (present) in `rm_dv` to set (deleted) in the `add_dv`. Once again, all
+            //  other rows are either cancelled in set subtraction, or were not set in either treemap.
+            //  Applying this to our deletion vectors:
+            //  add_dv - rm_dv =
+            //      Treemap_d(1, 2)
+            //    - Treemap_d(0, 1)
+            //    = Treemap_s(2)
+            //  The selection treemap shows that row 2 was deleted
+            //
+            //  # Conversion to Selection Vector
+            //  The selection treemap is converted to a selection vector by setting the bits:
+            //      Treemap_s(0) => [true]
+            //      Treemap_s(2) => [false, false, true]
+            //  All other rows are unselected (false).
+            let adds = &rm_dv - &add_dv;
+            let removes = add_dv - rm_dv;
+
+            let adds = (!adds.is_empty()).then_some(adds);
+            let removes = (!removes.is_empty()).then_some(removes);
+            (adds, removes)
+        }
+        (add_dv, None, CdfScanFileType::Add | CdfScanFileType::Cdc) => {
+            (Some(add_dv.unwrap_or_else(Default::default)), None)
+        }
+        (rm_dv, None, CdfScanFileType::Remove) => {
+            (None, Some(rm_dv.unwrap_or_else(Default::default)))
+        }
+    };
+    let treemap_to_bools = if scan_file.remove_dv.is_some() {
+        selection_treemap_to_bools
+    } else {
+        deletion_treemap_to_bools
+    };
+
+    let rm_scan_file = CdfScanFile {
+        scan_type: CdfScanFileType::Remove,
+        ..scan_file.clone()
```

Comment:
nit: This only gets used if we have Some `rm_dv`? Maybe rust is smart enough to delay creating it until we actually use it?

Reply:
Also, the add/rm paths below are the ~same code duplicated. Maybe we could solve both issues in one swoop with a helper method?
```rust
let resolve = |dv, scan_file_fn| dv.map(treemap_to_bools).map(|sv| ResolvedCdfScanFile {
    scan_file: scan_file_fn(),
    sselection_vector: (!sv.is_emtpy()).then_some(sv),
});
let removes = resolve(rm_dv, || CdfScanFile { ... });
let adds = resolve(add_dv, || scan_file);
```

Reply:
Hmm rust has a tricky type system. One of them is `Fn() -> CdfScanFile` while the other one is FnOnce. So I have to do a `clone()` in the `||scan_file.clone()` closure to make it Fn. I changed it to the resolve fn with a clone, but lmk if I should revert it.

Also the clone may be optimized by the compiler, but I'm not sure.

If you have tips on viewing optimizations for rust, please let me know! I'm familiar with GodBolt, but I feel like assembly is too low level for the optimizations to be clear.

Reply:
Oh, you probably just needed to do `move || scan_file`?

Reply:
Yeah the problem is that `move || scan_file` is `FnOnce`, while the other one is `Fn`. `move | | scan_file` is  `FnOnce` because we can only return the captured owned value once. Meanwhile `|| CdfScanFile { ... }` clones its reference to `scan_file` so it can be called repeatedly. Thus it's `Fn`

We could make the closure take `scan_file_fn: &dyn Fn() -> CdfScanFile`, but then we'd have to use `| | scan_file.clone()` for the add case.
```rust
    let resolve = |dv: Option<RoaringTreemap>, scan_file_fn: &dyn Fn() -> CdfScanFile| {
        dv.map(treemap_to_bools).map(|sv| ResolvedCdfScanFile {
            scan_file: scan_file_fn(),
            selection_vector: (!sv.is_empty()).then_some(sv),
        })
    };
    let removes = resolve(rm_dv, &|| CdfScanFile {
        scan_type: CdfScanFileType::Remove,
        ..scan_file.clone()
    });
    let adds = resolve(add_dv, &|| scan_file.clone());
```

I tried `scan_file_fn: &dyn FnOnce() -> CdfScanFile` but the compiler isn't happy because you need to consume ownership of the function to call a `FnOnce`
```rust
    let resolve = |dv: Option<RoaringTreemap>, scan_file_fn: &dyn FnOnce() -> CdfScanFile| {
        dv.map(treemap_to_bools).map(|sv| ResolvedCdfScanFile {
            scan_file: scan_file_fn(),
            selection_vector: (!sv.is_empty()).then_some(sv),
        })
    };

    let removes = resolve(rm_dv, &|| CdfScanFile {
        scan_type: CdfScanFileType::Remove,
        ..scan_file.clone()
    });
    let adds = resolve(add_dv, &|| scan_file);
    Ok([removes, adds].into_iter().flatten())
```

Moreover generics don't work with closures, so we'd need a proper function `fn resolve<T>(dv: ..., scan_file_fn: T)`,

Perhaps there's a way to specify generic trait impls so that  we monomorphize into two resolves: `resolve::<impl FnOnce>`, and `resolve::<impl Fn>` but by this point I felt like this added more lines and complexity than it was worth.